File: src/lib/proactiveContextHints.ts
import type { AgentRunContextHints } from "@/lib/agentRuns";

const REPO_ROOT_MARKERS = ["/src/", "/server/", "/lib/", "/app/", "/components/", "/pages/", "/routes/"];

const MAX_FOCUS = 10;
const MAX_HUB = 8;
const MAX_ENTRY = 6;
const MAX_TECH = 8;

export type ProactiveManifestLike = {
  scenes?: Array<{ file_path?: string | null }>;
  evidence_bundle?: {
    important_files?: string[];
    hub_files?: string[];
    entry_candidates?: string[];
    snippet_catalog?: unknown[];
    repo_stats?: { key_technologies?: string[] };
  } | null;
  knowledge_graph?: { summary?: { architecture?: string | null } } | null;
};

export type ProactiveGraphLike = {
  summary?: {
    hubFiles?: string[];
    entryPoints?: string[];
    keyTechnologies?: string[];
    architecturePattern?: string | null;
  } | null;
} | null;

export function normalizeContextPath(path: string): string | null {
  let cleaned = path.trim().replace(/\\/g, "/");
  if (!cleaned || cleaned.split("/").some((segment) => segment === "..")) return null;
  if (cleaned.startsWith("local://")) cleaned = cleaned.slice("local://".length).replace(/^\/+/, "");
  const lowered = cleaned.toLowerCase();
  if (lowered.startsWith("http://") || lowered.startsWith("https://")) return null;

  if (!cleaned.startsWith("/")) return cleaned.replace(/^\/+/, "");

  let best: string | null = null;
  for (const marker of REPO_ROOT_MARKERS) {
    const idx = lowered.lastIndexOf(marker);
    if (idx < 0) continue;
    const candidate = cleaned.slice(idx + 1);
    if (!best || candidate.length > best.length) best = candidate;
  }
  if (best) return best;
  return cleaned.replace(/^\/+/, "");
}

function normalizePathList(values: unknown, limit: number): string[] {
  if (!Array.isArray(values)) return [];
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const raw of values) {
    const rel = normalizeContextPath(String(raw ?? ""));
    if (!rel || seen.has(rel)) continue;
    seen.add(rel);
    ordered.push(rel);
    if (ordered.length >= limit) break;
  }
  return ordered;
}

function normalizeStringList(values: unknown, limit: number): string[] {
  if (!Array.isArray(values)) return [];
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const raw of values) {
    const text = String(raw ?? "").trim();
    if (!text || seen.has(text)) continue;
    seen.add(text);
    ordered.push(text.slice(0, 120));
    if (ordered.length >= limit) break;
  }
  return ordered;
}

function normalizeCount(value: unknown): number {
  const parsed = Number.parseInt(String(value ?? ""), 10);
  if (!Number.isFinite(parsed) || parsed < 0) return 0;
  return Math.min(parsed, 10_000);
}

export function buildProactiveContextHints(
  manifest: ProactiveManifestLike | null,
  graphData: ProactiveGraphLike,
): AgentRunContextHints {
  const focusFiles = normalizePathList(
    [
      ...(manifest?.scenes?.map((scene) => scene.file_path) ?? []),
      ...(manifest?.evidence_bundle?.important_files ?? []),
    ],
    MAX_FOCUS,
  );

  const hubFiles = normalizePathList(
    [...(manifest?.evidence_bundle?.hub_files ?? []), ...(graphData?.summary?.hubFiles ?? [])],
    MAX_HUB,
  );

  const entryFiles = normalizePathList(
    [...(manifest?.evidence_bundle?.entry_candidates ?? []), ...(graphData?.summary?.entryPoints ?? [])],
    MAX_ENTRY,
  );

  const technologies = normalizeStringList(
    [
      ...(manifest?.evidence_bundle?.repo_stats?.key_technologies ?? []),
      ...(graphData?.summary?.keyTechnologies ?? []),
    ],
    MAX_TECH,
  );

  const snippetCount = manifest?.evidence_bundle?.snippet_catalog?.length ?? 0;

  return {
    focusFiles,
    hubFiles,
    entryFiles,
    technologies,
    architecture:
      manifest?.knowledge_graph?.summary?.architecture ??
      graphData?.summary?.architecturePattern ??
      null,
    evidenceCount: snippetCount,
    snippetCount,
  };
}

export function serializeProactiveContextHints(
  hints: AgentRunContextHints | Record<string, unknown> | null | undefined,
): AgentRunContextHints {
  const raw = hints ?? {};
  const evidenceCount = normalizeCount(
    "evidenceCount" in raw ? raw.evidenceCount : "snippetCount" in raw ? raw.snippetCount : 0,
  );
  const snippetCount = normalizeCount(
    "snippetCount" in raw ? raw.snippetCount : "evidenceCount" in raw ? raw.evidenceCount : evidenceCount,
  );
  const architecture =
    typeof raw.architecture === "string" && raw.architecture.trim() ? raw.architecture.trim().slice(0, 240) : null;

  return {
    focusFiles: normalizePathList("focusFiles" in raw ? raw.focusFiles : [], MAX_FOCUS),
    hubFiles: normalizePathList("hubFiles" in raw ? raw.hubFiles : [], MAX_HUB),
    entryFiles: normalizePathList("entryFiles" in raw ? raw.entryFiles : [], MAX_ENTRY),
    technologies: normalizeStringList("technologies" in raw ? raw.technologies : [], MAX_TECH),
    architecture,
    evidenceCount,
    snippetCount,
  };
}

File: server/proactive_context_hints.py
from __future__ import annotations

from pathlib import Path
from typing import Any, Optional

MAX_FOCUS_FILES = 10
MAX_HUB_FILES = 8
MAX_ENTRY_FILES = 6
MAX_TECHNOLOGIES = 8
MAX_ARCHITECTURE_LEN = 240
MAX_EVIDENCE_COUNT = 10_000

REPO_ROOT_MARKERS = ("/src/", "/server/", "/lib/", "/app/", "/components/", "/pages/", "/routes/")

CONTEXT_HINT_FOCUS_BONUS = 0.08
CONTEXT_HINT_HUB_BONUS = 0.04
CONTEXT_HINT_ENTRY_BONUS = 0.04
CONTEXT_HINT_BONUS_CAP = 0.12


def normalize_repo_relative_path(path: Any) -> Optional[str]:
    cleaned = str(path or "").strip().replace("\\", "/")
    if not cleaned:
        return None
    if ".." in cleaned.split("/"):
        return None
    if cleaned.startswith("local://"):
        cleaned = cleaned[len("local://") :].lstrip("/")
    lowered = cleaned.lower()
    if lowered.startswith("http://") or lowered.startswith("https://"):
        return None

    if not cleaned.startswith("/"):
        return cleaned.lstrip("/")

    best: Optional[str] = None
    for marker in REPO_ROOT_MARKERS:
        idx = lowered.rfind(marker)
        if idx < 0:
            continue
        candidate = cleaned[idx + 1 :].lstrip("/")
        if best is None or len(candidate) > len(best):
            best = candidate
    if best:
        return best

    return cleaned.lstrip("/")


def _normalize_path_list(values: Any, limit: int, workspace: Optional[Path]) -> list[str]:
    if not isinstance(values, list):
        return []
    seen: set[str] = set()
    ordered: list[str] = []
    for raw in values:
        rel = normalize_repo_relative_path(raw)
        if not rel or rel in seen:
            continue
        if workspace is not None:
            try:
                absolute = (workspace / rel).resolve()
                workspace_root = workspace.resolve()
                if not str(absolute).startswith(str(workspace_root)):
                    continue
            except OSError:
                pass
        seen.add(rel)
        ordered.append(rel)
        if len(ordered) >= limit:
            break
    return ordered


def _normalize_string_list(values: Any, limit: int) -> list[str]:
    if not isinstance(values, list):
        return []
    seen: set[str] = set()
    ordered: list[str] = []
    for raw in values:
        text = str(raw or "").strip()
        if not text or text in seen:
            continue
        seen.add(text)
        ordered.append(text[:120])
        if len(ordered) >= limit:
            break
    return ordered


def _normalize_count(value: Any) -> int:
    try:
        count = int(value)
    except (TypeError, ValueError):
        return 0
    return max(0, min(count, MAX_EVIDENCE_COUNT))


def _normalize_architecture(value: Any) -> Optional[str]:
    text = str(value or "").strip()
    if not text:
        return None
    return text[:MAX_ARCHITECTURE_LEN]


def normalize_proactive_context_hints(
    raw: Any,
    *,
    workspace: Optional[Path] = None,
) -> dict[str, Any]:
    data = raw if isinstance(raw, dict) else {}
    evidence = _normalize_count(data.get("evidenceCount") or data.get("snippetCount"))
    snippet = _normalize_count(data.get("snippetCount") or data.get("evidenceCount") or evidence)
    return {
        "focusFiles": _normalize_path_list(data.get("focusFiles"), MAX_FOCUS_FILES, workspace),
        "hubFiles": _normalize_path_list(data.get("hubFiles"), MAX_HUB_FILES, workspace),
        "entryFiles": _normalize_path_list(data.get("entryFiles"), MAX_ENTRY_FILES, workspace),
        "technologies": _normalize_string_list(data.get("technologies"), MAX_TECHNOLOGIES),
        "architecture": _normalize_architecture(data.get("architecture")),
        "evidenceCount": evidence,
        "snippetCount": snippet,
    }


def context_hint_sets(hints: dict[str, Any]) -> tuple[set[str], set[str], set[str], set[str]]:
    focus = set(hints.get("focusFiles") or [])
    hub = set(hints.get("hubFiles") or [])
    entry = set(hints.get("entryFiles") or [])
    return focus, hub, entry, focus | hub | entry


def path_matches_hint(repo_relative: str, hint_paths: set[str]) -> bool:
    norm = normalize_repo_relative_path(repo_relative) or str(repo_relative or "").strip().lstrip("/")
    if not norm:
        return False
    if norm in hint_paths:
        return True
    for hint in hint_paths:
        if norm == hint or norm.endswith(f"/{hint}") or hint.endswith(f"/{norm}"):
            return True
    return False


def context_hint_flags_for_path(path: str, hints: dict[str, Any]) -> dict[str, bool]:
    focus, hub, entry, _ = context_hint_sets(hints)
    return {
        "focus": path_matches_hint(path, focus),
        "hub": path_matches_hint(path, hub),
        "entry": path_matches_hint(path, entry),
    }


def compute_context_hint_bonus(
    *,
    in_focus: bool = False,
    in_hub: bool = False,
    in_entry: bool = False,
) -> float:
    bonus = 0.0
    if in_focus:
        bonus += CONTEXT_HINT_FOCUS_BONUS
    if in_hub:
        bonus += CONTEXT_HINT_HUB_BONUS
    if in_entry:
        bonus += CONTEXT_HINT_ENTRY_BONUS
    return min(CONTEXT_HINT_BONUS_CAP, bonus)


def manifest_evidence_floor(local_count: int, hints: dict[str, Any], in_focus: bool) -> int:
    if not in_focus:
        return local_count
    manifest_count = _normalize_count(hints.get("evidenceCount") or hints.get("snippetCount"))
    if manifest_count <= 0:
        return local_count
    boosted = max(local_count, min(8, manifest_count // 4))
    return boosted


def merge_run_context_hints(candidate_path: str, dispatch_hints: Optional[dict[str, Any]]) -> dict[str, Any]:
    normalized = normalize_proactive_context_hints(dispatch_hints or {})
    focus_files = list(
        dict.fromkeys(
            [
                normalize_repo_relative_path(candidate_path) or candidate_path,
                *(normalized.get("focusFiles") or []),
            ],
        ),
    )[:MAX_FOCUS_FILES]
    return {
        "focusFiles": focus_files,
        "hubFiles": list(normalized.get("hubFiles") or []),
        "entryFiles": list(normalized.get("entryFiles") or []),
        "technologies": list(normalized.get("technologies") or []),
        "architecture": normalized.get("architecture"),
        "evidenceCount": int(normalized.get("evidenceCount") or 0),
        "snippetCount": int(normalized.get("snippetCount") or normalized.get("evidenceCount") or 0),
    }
